### PR TITLE
Production Release: TEM-204, TEM-188, TEM-199, TEM-202, TEM-203, TEM-104, TEM-192, TEM-205, TEM-115

### PR DESCRIPTION
**Pull request by postables**
## :construction_worker: Purpose
Production release, after the following 6 tasks everything should be ready
~TODO: remove private key entry from database when its removed~
~TODO: Pass jwt through queue system for allowing private network api access~
~TODO: add usage model integration~
~TODO: add client-based certificate auth to rabbitmq~
~TODO: cleanup ETH+RTC payment system~
TODO: refactor IPNS charging

### Dependencies

* IPFS Cluster is now at 0.8.0
* minio is removed
~* stripe-go has been added~

### Functionality

* When exporting private keys, they are removed from krab and the database
* When creating a user via cli, grant credits
* Optional decryption during download re-added
* you can now pay for credits with stripe!!!!
* email address *must not* contain `+` this is done to prevent abuse of the free account tier, and catch all routing within email systems. 
* better [api docs](https://gateway.temporal.cloud/ipns/docs.api.temporal.cloud/)
* email verification is now done without requiring any authentication, via a jwt. When signing up, a jwt for the user is generated, and a link is sent to the email address specified during registration. Clicking this link will then verify the jwt, parse the claims, and if everything checkouts email is activated.

### Optimizations

* Pinning is now streamlined, instead of first going through the queue system, pin on ipfs nodes, and then queue system to trigger a cluster pin. Now we send a message through rabbitmq to a random consumer which then triggers a cluster-wide pin
* `IpfsPinQueue` is now only used by private networks
* `rtfscluster` has been greatly reduced in size to only the needed calls
* There are nolonger two separate file upload calls, they have been condensed into one, allowing for encryption to be triggered based off whether or not the postform `passphrase` is provided

### Fixes

* Fixes logs all going into one file
* getObjectStat call no longer uses `key` it uses `hash` 
* all api calls are commented
* pin exchange is removed
* IPFS API Clients now honor the timeout variable
* request id properly populated in log files
* mail tests enabled through use of encrypted sendgrid key part of travis build
* fixes panics happening when shutting down the api

## :warning: Breaking Changes

* getObjectStat call
* cluster api call no longer exists
* advanced upload calls no longer exist
* getDepositAddress call no longer exists
* email token get route is removed
* email verification
* ipns publishing on private networks through the api is removed

## Pre-Merge Tasks (after review)

* within all RTrade repositories (non Nexus and non Lens) pin to version 2.0.0
* Update Temporal dependencies to v2.0.0 of all RTrade projects

## Post-Merge Tasks

* Branch master off of V2
* Change default branch to master
* Branch dev off of master
* Delete V2

## Notes:

Why do we remove keys once deleted?
* Because IPFS keys are primarily used to publish IPNS records, having multiple copies of keys used by independent sources can cause some nodes in the network to have non up-to-date records, which could cause conflictions.

Pin workflow changes:
* previously the pinning workflow was `pin api call -> rabbitmq fanout exchange publish to all consumers -> pin -> rabbitmq publish to cluster pin consumer -> pin to cluster` this was ineffective for a few reasons, as if a failure occurred on one node, or even both nodes, the users credit balance would be improper, and the users data usage would be improper. 
* The new workflow is `pin api call -> rabbitmq publish to a random cluster pin consumer -> cluster pin is triggered -> all nodes in cluster start pinning` This is more effective, and allows for better handling of  failed pins. Furthermore, it is now trivial to trigger a re-pin as we can leverage cluster state operations.

What's changed with private network billing:
* outside of the development environment, we will not charge data usage for private networks, and instead have a monthly fee which is resposible for private network node resource allocation, and bulk storage allotment. With the public API, usage is accounted for everything (keys, pubsub, pin, ipns, etc..) but none of these will apply towards private networks. So you're free to do as you please with your private networks.

Keys created do not apply to private networks:
* Keys created are not usable with private networks. Due to our modified keystore, which is inaccessible with private networks, any keys you create are not usable in private networks. Instead, private networks rely on the built-in keystore present with all IPFS nodes.

file uploads have changed:
* file uploads have now been condensed into a single call, as opposed to previously where there was a regular file add, and advanced file add. Previously the advanced upload was just used for encryption. By condensing the two, we now reduce the total amount of queues by 1, and remove a big dependency (minio) from our dependency tree. Additionally, encrypted uploads now immediately return the hash of the file.

why was private network ipns publishing removed:
* because we leverage krab as our keystore, all keys generated for now are unable to be injected into private ipfs network nodes. That being said, if users access their private network API directly through `delegator` they can create keys, and publish IPNS records. This removal *only* applies to IPNS record publishing through Temporal's API.

stripe:
* the stripe implementation was somewhat hacky and required dangerously injecting html code into the web interface. This has been removed so that the bulk of stripe can be implemented via the web interface.

### CODE REVIEW

**Comment by postables** [comment]
Figured out a problem with our usage model.

Take the following program which emits gigabyte size in varying units

```Golang
package main

import (
	"fmt"

	"github.com/c2h5oh/datasize"
)

func main() {
	fmt.Println("datasize gbytes", datasize.GB.GBytes())
	fmt.Println("datasize bytes", datasize.GB.Bytes())
}
```

The output is
```
datasize gbytes 1
datasize bytes 1073741824
```

The problem is that we are tracking the `gbytes` within our usage model which is causing failures. We need to track the `bytes` since ipfs provides file sizes as bytes, and when uploading files the `size` from the file handler is in bytes

File: Makefile [hunk 1]
```diff
@@ -55,9 +55,6 @@ testenv:
 	go run cmd/temporal/main.go -config ./testenv/config.json --db.no_ssl migrate
 	make api-user
 	make api-admin
-	# create minio bucket
-	make bucket
-	make bucket BUCKET=test
```
Comment (bobheadxi) [inline Makefile]: i didnt see any mention of buckets in the PR desc, are we dropping minio usage completely?

File: api/v2/api_test.go [hunk 2]
```diff
@@ -663,7 +664,7 @@ func TestAPI_HandleQueuError_Failure(t *testing.T) {
 	fakeLens := &mocks.FakeIndexerAPIClient{}
 	fakeOrch := &mocks.FakeServiceClient{}
 	fakeSigner := &mocks.FakeSignerClient{}
-	api, err := Initialize(cfg, "", true, logger, fakeLens, fakeOrch, fakeSigner)
+	api, err := Initialize(cfg, "", true, true, logger, fakeLens, fakeOrch, fakeSigner)
```
Comment (bobheadxi) [inline api/v2/api_test.go]: why do we have `debug` AND `devMode` as params? to me they imply the same thing

if we need it, I would rather a option struct be introduced:

```go
type Options struct {
   DebugLogs bool
   DevMode bool
}
```

File: api/v2/context.go [hunk 3]
```diff
@@ -63,6 +64,14 @@ func GetAuthenticatedUserFromContext(c *gin.Context) (string, error) {
 	return strID, nil
 }
 
+// GetAuthToken is used to retrieve the jwt token
+// from an authenticated request
+func GetAuthToken(c *gin.Context) string {
+	token := strings.Split(c.GetHeader("Authorization"), " ")[1]
+	fmt.Println(token)
```
Comment (bobheadxi) [inline api/v2/context.go]:
```suggestion
```

File: api/v2/context.go [hunk 4]
```diff
@@ -63,6 +64,14 @@ func GetAuthenticatedUserFromContext(c *gin.Context) (string, error) {
 	return strID, nil
 }
 
+// GetAuthToken is used to retrieve the jwt token
+// from an authenticated request
+func GetAuthToken(c *gin.Context) string {
+	token := strings.Split(c.GetHeader("Authorization"), " ")[1]
```
Comment (bobheadxi) [inline api/v2/context.go]: should check for length first

File: .travis.yml [hunk 5]
```diff
@@ -5,7 +5,11 @@ services:
 - docker
 sudo: required
 env:
-- SSL_MODE_DISABLE=true
+  matrix:
+  - SSL_MODE_DISABLE=true
+  global:
+  # this decodes to env variable SENDGRID_API_KEY
+  - secure: GXKJ+1Wa+pqUm1PKHJMbDrejSfIsqhsZ6l88i0paTvVou9T/mv03mx9hzrSLsYXFFVjinfbKysxMtEorTFDl0YBSNxuPc7eaiJinnoptubBh++bvEzF/A8wmWHgIBxroFIrK4SNsnreCXPqT1YBQZWi3JenMkzS68j1oa/uPa5ODZ3rjx1Wmu6hSANSZPqzOoGO6lnRP8G30oiFzbawEnB/52iBcHhPtffUihLFnn3k1wXyL8fpi4JYuyNQe84Br+w91KSS2nMmG+RVvJWumHNGXFyjEVv0n7HbirjCBI1iZo50bTdFtujkGSOCoHqM1hwa2yLWISipkb19Ls4eeWcWZBFpcdBhIIeHEmz9iCXxn0ksIGzuw00Xof/HTcWUpnzzQFq4E8iVE4tmMrFbYEcD5vcGB3S7bURjTD38uZ3/7I4Oyuo33/WQGdDNIzugU/dycGkSNzwTeIB3yJfpr/wbC6lU5RnADh5Ej5hGreAfkng/iuC9T36Fnn9u4fTocj++smOxZSjbbKsyekNNdadYmU0sfd6Ka9mFcv8H6sDs7zwqyZLTDu6P57CN5VFz4Yjc1emupFX5NFWor5jull9cr+ilOmfnBSCQPV0ArKrpKjrL9K9NJAhRKOqxmpLKx3pozGnrQMFKM0xsSuTcOCdyTdZF8JEQESeENHzwyoB4=
```
[the remark this answers is not in the document]
Comment (postables) [reply .travis.yml]: I'm not quite sure I was going off the travis [docs](https://docs.travis-ci.com/user/environment-variables/) and when running with `travis encrypt .... --add` it populated the fields like this

File: api/v2/routes_rtfs_test.go [hunk 6]
```diff
@@ -157,7 +149,21 @@ func Test_API_Routes_IPFS_Public(t *testing.T) {
...
+	); err != nil {
+		t.Fatal(err)
+	}
+	if apiResp.Code != 400 {
+		t.Fatal("bad response status code from /v2/ipfs/public/pubsub/publish")
+	}
+
+	// test object stat (success)
 	// /v2/ipfs/stat
 	var interfaceAPIResp interfaceAPIResponse
```
[the remark this answers is not in the document]
Comment (postables) [reply api/v2/routes_rtfs_test.go]: not sure what you mean? `interfaceAPIResponse` is used when calls from the API return a response that I couldn't find into a `map[string]string` response. Object stat for example returns a response of

```Golang
type ObjectStats struct {
	Hash           string
	BlockSize      int
	CumulativeSize int
	DataSize       int
	LinksSize      int
	NumLinks       int
}
```

File: .travis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (postables) [reply .travis.yml]: Looks like we can, good suggestion :ok_hand:

File: api/v2/routes_rtfs_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (bobheadxi) [reply api/v2/routes_rtfs_test.go]: ah I see them all, looks like they're only used in tests which is fine 👍

**Comment by postables** [comment]
Update: Free tier accounts are restricted when making encrypted uploads. The max size for encrypted uploads for free users is 275MB, while non free users the max size is the max upload size (1GB)

**Comment by postables** [comment]
Update: disable private networks in production, only usable within development environment until we leave beta for Nexus.

**Pull request merged**